### PR TITLE
Dev flow

### PR DESCRIPTION
**Pull request by lustoykov**
Very much WIP and open to suggestions!

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,57 +1,97 @@
...
-## Gatsby
+`Install node and npm`
+
+### Gatsby
 
 `npm install -g gatsby-cli`
 
+## Context
+
+The gatsby frontend resides currently in the root directory. The intention is to move it to e `frontend/` folder.
```
Comment (mstockerl) [inline README.md]:
```suggestion
The gatsby frontend resides currently in the root directory. The intention is to move it to a `frontend/` folder.
```

File: README.md [hunk 2]
```diff
@@ -1,57 +1,97 @@
...
+
+## Setting Up
+
+1. Run `npm install` in the root dir
+
+## Development
+
+1. Run `npm run start`
+2. Any environemnt variables go into `.env.development` for local development and into `cloudbuild-prod|staging.yaml` for prod / staging. This is a bit inconsistent, it might be best to move those from `cloudbuild-prod|staging.yaml` in `.env.staging` and `.env.production`, but yolo unless you wanna fix it.
```
Comment (mstockerl) [inline README.md]: The documentation of Gatsby suggest not to bring `.env.*` into a repository.
https://www.gatsbyjs.org/docs/environment-variables/#example

I'd rather keep it "inconsistent" for now, as we have the option of encryption in the cloudbuild.yaml files.

File: README.md [hunk 3]
```diff
@@ -1,57 +1,97 @@
...
-Make sure you are in the root directory
-Start the development server using
-```
-gatsby develop
-```
-and visit
-```
-http://localhost:8000/
-```
+3. Now you can experiment against the functions that run locally and are accessible via localhost, but they access our staging database & storage.
```
Comment (mstockerl) [inline README.md]: What happens with Functions with a trigger like a new entry in the database? 
Do they run locally and in staging then? Or is only one triggered? 🤔

File: README.md [hunk 4]
```diff
@@ -1,57 +1,97 @@
...
+1. Write a cloud function
+2. Write unit tests
+3. Write integration tests
+4. Go manually over the flow to see if it works
+5. Deploy the function to staging & make sure it works
+6. Deploy the function to production
+
+## Deployment
+
+We don't have CI yet. You'll need to be logged into gcloud cli and have the appropriate premissions.
```
Comment (mstockerl) [inline README.md]: I'd like to update that. :)

File: functions/__integrationtests__/handleFormSubmit.test.js [hunk 5]
```diff
@@ -0,0 +1,71 @@
+jest.mock('cors');
+const testEnv = require('firebase-functions-test')(
+  {
+    projectId: 'dps-website-staging-0',
+  },
+  './dps-website-staging.firebase.key.json'
```
Comment (mstockerl) [inline functions/__integrationtests__/handleFormSubmit.test.js]: We need to integrate this into the build process. :) 
Do you want to do it or should I create a PR for it?

File: functions/src/handleFormSubmit.ts [hunk 6]
```diff
@@ -1,36 +1,118 @@
...
+    const busboy = new Busboy({ headers: req.headers });
+    const data: { [key: string]: string | Buffer | Promise<any> } = {};
+
+    // collect file stream and pipe them to the storage, return promises as values
+    busboy.on(
+      'file',
+      async (fieldname: string, readFileStream: any, filename: string) => {
+        const filePromise = new Promise(async (fileResolve, fileReject) => {
+          const bucket = STORAGE_BUCKET_URL;
+          const name = `participants/${Date.now()}_${filename}`;
```
Comment (mstockerl) [inline functions/src/handleFormSubmit.ts]: I'd like to have a similar path like we have in the database. 
So first a folder for each batch and one folder for each participant. In this one, we can store the CV and/or cover letter. 

This way, we can track more easily whether the CVs get deleted after the grace period.

File: functions/src/index.ts [hunk 7]
```diff
@@ -3,25 +3,32 @@ import * as TrelloNodeAPI from 'trello-node-api';
...
+  trello.setOauthToken(functions.config().trello.token);
+
+  exports.createTrelloCard = functions
+    .region(defaultRegion)
+    .firestore.document(`batches/{batch}/applications/{applicationId}`)
+    .onCreate(async (snap, context) => {
+      createTrelloCard.handler(snap, trello);
+    });
+} else {
+  console.error('Trello config unavailable');
```
Comment (mstockerl) [inline functions/src/index.ts]: Can we throw an error here instead?

I'd rather have a function not started and hence broken build than missing it and losing applications in trello.

**Comment by lustoykov** [comment]
Great comments & suggestions, thanks! let's finish this off on Monday! :-)

File: README.md (hunk 3 above)
Comment (lustoykov) [reply README.md]: I think if we want local triggers we have to use emulators. I guess for now staging only is the way to test or write tests

File: functions/__integrationtests__/handleFormSubmit.test.js (hunk 5 above)
Comment (lustoykov) [reply functions/__integrationtests__/handleFormSubmit.test.js]: YoU :D

File: functions/src/handleFormSubmit.ts (hunk 6 above)
Comment (lustoykov) [reply functions/src/handleFormSubmit.ts]: yep!

**Comment by lustoykov** [comment]
Needs re-review

**Comment by mstockerl** [comment]
👍

**Pull request merged**